containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java

```diff
@@ -70,6 +73,54 @@ public Status status() {
         }
     }
 
+    public enum CancellationPolicy {
+
+        CONTEXT() {
+            @Override
+            public void apply(
+                final Context.CancellableContext cancellableContext,
+                final ClientCallStreamObserver<?> observer,
+                final String message) {
+                cancellableContext.cancel(new RuntimeException(message));
+            }
+        },
+        EXCEPTION() {
+            @Override
+            public void apply(final Context.CancellableContext cancellableContext,
+                              final ClientCallStreamObserver<?> observer,
+                              final String message) {
+                throw new CancellationException(message);
+            }
+        },
+        NONE() {
+            @Override
+            public void apply(final Context.CancellableContext cancellableContext,
+                              final ClientCallStreamObserver<?> observer,
+                              final String message) {
+                // do nothing
+            }
+        },
+        OBSERVER() {
+            @Override
+            public void apply(final Context.CancellableContext cancellableContext,
+                              final ClientCallStreamObserver<?> observer,
+                              final String message) {
+                if (observer == null) {
+                    throw new IllegalStateException("Cannot cancel when stream observer is missing!");
+                } else {
+                    final StatusRuntimeException error = Status.CANCELLED
+                        .withDescription("Observer: " + message)
+                        .asRuntimeException();
+                    observer.cancel(message, error);
+                }
+            }
+        };
+
+        public abstract void apply(final Context.CancellableContext cancellableContext,
```

Comment:
Nitpick: Can we avoid requiring parameters for implementations that don't care about the various arguments?

Reply:
Skipped, as if I continue with the enum solution, then will need casting in the individual methods.